### PR TITLE
Add driver for accelerometer

### PR DESCRIPTION
**Pull request by VictorDuma-zz**

### CODE REVIEW

File: inc/MMA8453.h [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+#define MMA8453_H
+
+#include "CodalConfig.h"
+#include "CodalComponent.h"
+#include "CodalUtil.h"
+#include "CoordinateSystem.h"
+#include "Pin.h"
+#include "MbedI2C.h"
+#include "Accelerometer.h"
+#include "BrainPadIO.h"
```
Comment (Arke64) [inline inc/MMA8453.h]: Verify these headers are all needed, particularly BrainPadIO.h

File: inc/MMA8453.h [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+#define MMA8453_DEFAULT_ADDR    0x38
+#define MMA8453_WHOAMI_VAL      0x3A
+
+//Register map (partial)
```
Comment (Arke64) [inline inc/MMA8453.h]: Remove this comment

File: inc/MMA8453.h [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+#define STATUS_REG       0x00
+#define STATUS_REG_AUX   0x07
+#define OUT_X_MSB        0x01
+#define OUT_X_LSB        0x02
+#define OUT_Y_MSB        0x03
+#define OUT_Y_LSB        0x04
+#define OUT_Z_MSB        0x05
+#define OUT_Z_LSB        0x06
+#define WHOAMI           0x0D
```
Comment (Arke64) [inline inc/MMA8453.h]: Needs underscores to match the rest

File: inc/MMA8453.h [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+#define HP_FILTER_CUTOFF 0x0F
+#define FT_MT_THS        0x17
+#define FF_MT_COUNT      0x18
+#define FF_MT_CFG        0x15
+#define XYZ_DATA_CFG     0x0E
+
+namespace codal {
+    class MMA8453 : public Accelerometer {
+        codal::_mbed::I2C i2c;  
+        Pin             &int1;              // Data ready interrupt.
```
Comment (Arke64) [inline inc/MMA8453.h]: & should be on the type, not the name. Further, is this member used anywhere?

File: inc/MMA8453.h [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        uint16_t        address;            // I2C address of this accelerometer.
+        uint16_t        samplePeriod;       // The time between samples, in milliseconds.
+        uint8_t         sampleRange;        // The sample range of the accelerometer in g.
+        Sample3D        sample;             // The last sample read.
+
+    public:
+        MMA8453(Pin& sda, Pin& scl, Pin &_int1, CoordinateSpace &coordinateSpace, uint16_t address = MMA8453_DEFAULT_ADDR, uint16_t id = DEVICE_ID_ACCELEROMETER);
```
Comment (Arke64) [inline inc/MMA8453.h]: Same & style issue

File: inc/MMA8453.h [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+#define FF_MT_COUNT      0x18
+#define FF_MT_CFG        0x15
+#define XYZ_DATA_CFG     0x0E
+
+namespace codal {
+    class MMA8453 : public Accelerometer {
+        codal::_mbed::I2C i2c;  
+        Pin             &int1;              // Data ready interrupt.
+        uint16_t        address;            // I2C address of this accelerometer.
+        uint16_t        samplePeriod;       // The time between samples, in milliseconds.
```
Comment (Arke64) [inline inc/MMA8453.h]: Many of these members appear to be duplicated from the base class. We should just use the base class version, being sure to call the base constructor.

File: inc/MMA8453.h [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        uint8_t         sampleRange;        // The sample range of the accelerometer in g.
+        Sample3D        sample;             // The last sample read.
+
+    public:
+        MMA8453(Pin& sda, Pin& scl, Pin &_int1, CoordinateSpace &coordinateSpace, uint16_t address = MMA8453_DEFAULT_ADDR, uint16_t id = DEVICE_ID_ACCELEROMETER);
+
+        int setPeriod(int period);
```
Comment (Arke64) [inline inc/MMA8453.h]: Do we need to even override get period and range? Does the base class handle that for us? Can we just override set, do any logic we may need, then call the base's set?

File: inc/MMA8453.h [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        virtual int getPeriod();
+        int setRange(int range);
+        virtual int getRange();
+        int whoAmI();
```
Comment (Arke64) [inline inc/MMA8453.h]: Is this needed?

File: inc/MMA8453.h [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+        Sample3D getSample();
+        void WriteRegister(uint8_t, uint8_t);
```
Comment (Arke64) [inline inc/MMA8453.h]: Should be private. Needs proper casing and parameter names.

File: inc/MMA8453.h [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    public:
+        MMA8453(Pin& sda, Pin& scl, Pin &_int1, CoordinateSpace &coordinateSpace, uint16_t address = MMA8453_DEFAULT_ADDR, uint16_t id = DEVICE_ID_ACCELEROMETER);
+
+        int setPeriod(int period);
+        virtual int getPeriod();
+        int setRange(int range);
+        virtual int getRange();
+        int whoAmI();
+
+        Sample3D getSample();
```
Comment (Arke64) [inline inc/MMA8453.h]: Is this needed?

File: model/BrainPad.cpp [hunk 11]
```diff
@@ -44,7 +44,9 @@ BrainPad::BrainPad() :
     buttonUp(io.buttonUp, ID_PIN_BUTTON_UP, DEVICE_BUTTON_ALL_EVENTS, ACTIVE_LOW, PullMode::Up),
     buttonDown(io.buttonDown, ID_PIN_BUTTON_DOWN, DEVICE_BUTTON_ALL_EVENTS, ACTIVE_LOW, PullMode::Up),
     buttonLeft(io.buttonLeft, ID_PIN_BUTTON_LEFT, DEVICE_BUTTON_ALL_EVENTS, ACTIVE_LOW, PullMode::Up),
-    buttonRight(io.buttonRight, ID_PIN_BUTTON_RIGHT, DEVICE_BUTTON_ALL_EVENTS, ACTIVE_LOW, PullMode::Up)
+    buttonRight(io.buttonRight, ID_PIN_BUTTON_RIGHT, DEVICE_BUTTON_ALL_EVENTS, ACTIVE_LOW, PullMode::Up),
+    space(CoordinateSystem::NORTH_EAST_UP, true, COORDINATE_SPACE_ROTATED_0),
```
Comment (Arke64) [inline model/BrainPad.cpp:48]: Do we need to store the coordinate system on the BrainPad? Can it be constructed in-place in the accelerometer initializer?

File: model/BrainPad.h [hunk 12]
```diff
@@ -47,6 +47,8 @@ DEALINGS IN THE SOFTWARE.
 #include "CodalFiber.h"
 #include "MessageBus.h"
 #include "BrainPadDisplay.h"
+#include "MMA8453.h"
+
```
Comment (Arke64) [inline model/BrainPad.h]: Remove this added newline

File: samples/main.cpp [hunk 13]
```diff
@@ -210,6 +210,42 @@ void TestServo() {
     brain.sleep(200);
 }
 
+void TestAccelerometer() {
+    int xAxis = brain.accelerometer.getX();
```
Comment (Arke64) [inline samples/main.cpp]: Should be named x, y, z

File: samples/main.cpp [hunk 14]
```diff
@@ -210,6 +210,42 @@ void TestServo() {
...
+    int yAxis = brain.accelerometer.getY();
+    int zAxis = brain.accelerometer.getZ();
+
+    if (brain.accelerometer.getGesture() == ACCELEROMETER_EVT_FACE_DOWN)
```
Comment (Arke64) [inline samples/main.cpp]: Use a switch, not an if. Set the gesture string in the switch cases, don't call DrawString itself.

```cs
const char* msg = nullptr;
switch (brain.acceleroemter.getGesture()) {
    case ACCELEROMETER_EVT_TILT_RIGHT: msg = "TILT_RIGHT"; break;
    ...
}

if (msg != nullptr)
    DrawString(1, 36, msg);
```

File: samples/main.cpp [hunk 15]
```diff
@@ -210,6 +210,42 @@ void TestServo() {
...
+    if (brain.accelerometer.getGesture() == ACCELEROMETER_EVT_FACE_UP)
+        DrawString(1, 36, "FACE UP   ");
+    if (brain.accelerometer.getGesture() == ACCELEROMETER_EVT_FREEFALL)
+        DrawString(1, 36, "FREEFALL  ");
+    if (brain.accelerometer.getGesture() == ACCELEROMETER_EVT_3G)
+        DrawString(1, 36, "EVT_3G    ");
+    if (brain.accelerometer.getGesture() == ACCELEROMETER_EVT_8G)
+        DrawString(1, 36, "EVT_8G    ");
+
+    std::string x = "X:" + std::to_string(xAxis) + "       ";
```
Comment (Arke64) [inline samples/main.cpp]: Why all the extra space at the end?

File: source/MMA8453.cpp [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+    {8, 0x02},
+};
+
+CREATE_KEY_VALUE_TABLE(accelerometerRange, accelerometerRangeData);
+
+MMA8453::MMA8453(Pin& sda, Pin& scl, Pin &_int1, CoordinateSpace &coordinateSpace, uint16_t address,  uint16_t id) : Accelerometer(coordinateSpace, id), i2c(sda, scl), int1(_int1), sample() {
+    this->id = id;
+    this->status = 0;
+    this->address = address;
+    // Update internal state at +/- 2g.
```
Comment (Arke64) [inline source/MMA8453.cpp]: Remove these comments

File: source/MMA8453.cpp [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+// Configuration table for available g force ranges.
+static const KeyValueTableEntry accelerometerRangeData[] = {
+    {2, 0x00},
+    {4, 0x01},
+    {8, 0x02},
+};
+
+CREATE_KEY_VALUE_TABLE(accelerometerRange, accelerometerRangeData);
+
+MMA8453::MMA8453(Pin& sda, Pin& scl, Pin &_int1, CoordinateSpace &coordinateSpace, uint16_t address,  uint16_t id) : Accelerometer(coordinateSpace, id), i2c(sda, scl), int1(_int1), sample() {
```
Comment (Arke64) [inline source/MMA8453.cpp]: Improper & style, extra space before last parameter.

File: source/MMA8453.cpp [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+    this->address = address;
+    // Update internal state at +/- 2g.
+    this->sampleRange = 2;
+    // Configure and enable the accelerometer.
+    configure();
+}
+
+int MMA8453::configure() {
+    uint8_t value;
+    WriteRegister(CTRL_REG1, 0x00);//Sleep mode
```
Comment (Arke64) [inline source/MMA8453.cpp]: Provide named constants for these values instead of comments.

File: source/MMA8453.cpp [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+    WriteRegister(CTRL_REG4, 0x01);//Enable Transient Detection Interrupt in the System. 
+    WriteRegister(CTRL_REG5, 0x01);//CTRL_REG5 register to determine the routing table for the INT1
+    WriteRegister(CTRL_REG1, 0x01);//Active mode
+
+    return DEVICE_OK;
+}
+
+void MMA8453::WriteRegister(uint8_t reg, uint8_t val) {
+    uint8_t writeData[2];
+    writeData[0]=reg;
```
Comment (Arke64) [inline source/MMA8453.cpp]: Need spacing around the `=`

File: source/MMA8453.cpp [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+    writeData[1]=val;
+
+    i2c.write(MMA8453_DEFAULT_ADDR, writeData, 2, 0); 
+}
+
+int MMA8453::whoAmI() {
+    uint8_t data;
+    int result;
+
+    result = i2c.readRegister(address, WHOAMI, &data, 1);
```
Comment (Arke64) [inline source/MMA8453.cpp]: Declare the variable inline with its assignment `int result = i2c....`

File: source/MMA8453.cpp [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+
+int MMA8453::updateSample() {
+    uint8_t readData[6];
+
+    //check new data with interrupt pin
+    if(int1.getDigitalValue() == 0) {
+        // read XYZ data
+        i2c.read(MMA8453_DEFAULT_ADDR, OUT_X_MSB, readData, 6);
+        //read MSB and LSB values
+        sample.x = readData[0] << 2 | readData[1] >> 6;
```
Comment (Arke64) [inline source/MMA8453.cpp]: Use parens for clarity `(readData[0] << 2) | (readData[1] >> 6)`

File: source/MMA8453.cpp [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+int MMA8453::requestUpdate() {
+    updateSample();
+
+    return DEVICE_OK;
+}
+
+int MMA8453::updateSample() {
+    uint8_t readData[6];
+
+    //check new data with interrupt pin
```
Comment (Arke64) [inline source/MMA8453.cpp]: Remove most of these comments

File: source/MMA8453.cpp [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+    if(int1.getDigitalValue() == 0) {
+        // read XYZ data
+        i2c.read(MMA8453_DEFAULT_ADDR, OUT_X_MSB, readData, 6);
+        //read MSB and LSB values
+        sample.x = readData[0] << 2 | readData[1] >> 6;
+        sample.y = readData[2] << 2 | readData[3] >> 6;
+        sample.z = readData[4] << 2 | readData[5] >> 6;
+
+        // scale data to milli g. 256 counts/g for 2 g
+        sample.x = (sample.x * 1000) / 256;
```
Comment (Arke64) [inline source/MMA8453.cpp]: We need to scale these values properly to the requested scale, 2g, 4g, or 8g.

File: source/MMA8453.cpp [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+        if(sample.x > 1024)
+            sample.x -= 4000;
```
Comment (Arke64) [inline source/MMA8453.cpp]: Shouldn't this be `-= 2048`?

**Review by Arke64 — CHANGES_REQUESTED** [review]

File: inc/MMA8453.h (hunk 4 above)
Comment (VictorDuma-zz) [reply inc/MMA8453.h]: updateSample() uses interrupt pin

Comment (Arke64) [reply inc/MMA8453.h]: Does it work as expected?

**Review by Arke64 — APPROVED** [review]

**Pull request merged**